changefeedccl: Ensure correct ordering.

miretskiy opened the pull request:
When async flushing enabled, the following sequence of events is possible (even if unlikely):
1. `k@t1` is emitted, causing async flush to write file `f1`
2. `k@t2` is emitted, causing async flush to write file `f2`
3. `f2` is written out before `f1`.

In this unlikely scenario -- and the reason why it's unlikely is that we have to generate megabytes of data to cause a flush, unless, file_size parameter was pathologically small -- if a client were to read the contents of the directory right after step 2, and then read directory after step 3, the client will first observe `k@t2`, and then observe `k@t1` -- which is a violation of ordering guarantees.

A simple fix to the above problem is to simply maintain the queue of pending flushes.  This approach is simple, it still has a reasonalbe throughput since Emit rows are not blocked. However it results in somewhat lower overall throughput since there is only 1 file being written out at any time; thus, the queue tends to build up, and pushback mechanism slows down the throughput.

Instead, this PR adopts an approach to allow throughput sensitive phases -- i.e. initial scan and schema change backfills -- to happen fully asynchronously, while ensuring that during normal operations, file reordering, as described above, cannot happen.

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89870)
<!-- Reviewable:end -->

miretskiy commented:
Going to test/benchmark simpler solution first.

Commit:
changefeedccl: Ensure correct file ordering.

When async flushing enabled, the following sequence of events is
possible (even if very unlikely):

k@t1 is emitted, causing async flush to write file f1
k@t2 is emitted, causing async flush to write file f2
f2 is written out before f1.

In this unlikely scenario -- and the reason why it's unlikely is that we
have to generate megabytes of data to cause a flush, unless, file_size
parameter was pathologically small -- if a client were to read the
contents of the directory right after step 2, and then read directory
after step 3, the client will first observe k@t2, and then observe k@t1
-- which is a violation of ordering guarantees.

This PR fixes this issue by adopting a queue so that if there
is a pending flush in flight, the next flush is queued behind.

It is possible that this simply approach may result in throughput
decrease.  This situation will occur if the underlying
storage (s3 or gcp) cannot keep up with writing out data before
the next file comes in.  However, at this point, such
situation is unlikely for two reasons: one, the rest of
changefeed machinery must be able to generate one or more files
worth of data before the previous flush completes -- and this
task in of itself is not trivial, and two, changefeed must
have enough memory allocated to it so that pushback mechanism
does not trigger.  The above assumption was validated in reasonably
sized test -- i.e. the non-zero queue depth was never observed.
Nonetheless, this PR also adds a log message which may be helpful
to detect the situation when the sink might not keep up with the
incoming data rate.

A more complex solution -- for example, allow unlimited inflight
requests during backfill -- may be revisited later, if the above
assumption proven incorrect.

Fixes #89683

Release note: None.
